tests/state_serde.rs

```diff
@@ -0,0 +1,115 @@
+//! State like [`ListState`], [`TableState`] and [`ScrollbarState`] can be serialized and
+//! deserialized through serde. This allows saving your entire state to disk when the user exits the
+//! the app, and restore it again upon re-opening the app.
+//! This way, they get right back to where they were, without having to re-seek to their previous
+//! position, if that's applicable for the app at hand.
+//!
+//! **Note**: For this pattern to work easily, you need to have some toplevel struct which stores
+//! _only_ state and not any draw commands.
+//!
+//! **Note**: For many applications, it might be beneficial to instead keep your own state and
+//! instead construct the state for widgets on the fly instead, if that allows you to express you
+//! the semantic meaning of your state better or only fetch part of a dataset.
+
+use ratatui::{backend::TestBackend, prelude::*, widgets::*};
+use serde::{Deserialize, Serialize};
+
+#[test]
```

Comment:
I found reading these tests to be a bit confusing and an idea about how to rewrite them so that the intent is clearer. Does this still capture your intent?

```rust
#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
struct AppState {
    list_state: ListState,
}

impl AppState {
    fn select_list_index(&mut self, index: usize) {
        self.list_state.select(Some(index));
    }
}

/// Renders the list to a TestBackend and asserts that the result matches the expected buffer.
#[track_caller]
fn assert_rendered_list_eq(list: List<'_>, list_state: &mut ListState, expected: &Buffer) {
    let backend = TestBackend::new(expected.area().width, expected.area().height);
    let mut terminal = Terminal::new(backend).unwrap();
    terminal
        .draw(|f| f.render_stateful_widget(list, f.size(), list_state))
        .unwrap();
    terminal.backend().assert_buffer(&expected);
}

#[test]
fn default_state() {
    let list = List::new([
        "awa", "banana", "Cats!!", "d20", "Echo", "Foxtrot", "Golf", "Hotel", "IwI", "Juliett",
    ])
    .highlight_symbol(">>");
    let mut state = AppState::default();

    let expected = Buffer::with_lines(vec![
        "awa       ",
        "banana    ",
        "Cats!!    ",
        "d20       ",
        "Echo      ",
    ]);
    assert_rendered_list_eq(list.clone(), &mut state.list_state, &expected);

    let serialized_state = serde_json::to_string_pretty(&state).unwrap();
    let mut deserialized_state: AppState = serde_json::from_str(&serialized_state).unwrap();

    assert_rendered_list_eq(list, &mut deserialized_state.list_state, &expected)
}

#[test]
fn selected_state() {
    let list = List::new([
        "awa", "banana", "Cats!!", "d20", "Echo", "Foxtrot", "Golf", "Hotel", "IwI", "Juliett",
    ])
    .highlight_symbol(">>");
    let mut state = AppState::default();
    state.select_list_index(1);

    let expected = Buffer::with_lines(vec![
        "  awa     ",
        ">>banana  ",
        "  Cats!!  ",
        "  d20     ",
        "  Echo    ",
    ]);
    assert_rendered_list_eq(list.clone(), &mut state.list_state, &expected);

    let serialized_state = serde_json::to_string_pretty(&state).unwrap();
    let mut deserialized_state: AppState = serde_json::from_str(&serialized_state).unwrap();

    assert_rendered_list_eq(list, &mut deserialized_state.list_state, &expected)
}

#[test]
fn scrolled_state() {
    let list = List::new([
        "awa", "banana", "Cats!!", "d20", "Echo", "Foxtrot", "Golf", "Hotel", "IwI", "Juliett",
    ])
    .highlight_symbol(">>");
    let mut state = AppState::default();
    state.select_list_index(8);

    let expected = Buffer::with_lines(vec![
        "  Echo    ",
        "  Foxtrot ",
        "  Golf    ",
        "  Hotel   ",
        ">>IwI     ",
    ]);

    assert_rendered_list_eq(list.clone(), &mut state.list_state, &expected);

    let serialized_state = serde_json::to_string_pretty(&state).unwrap();
    let mut deserialized_state: AppState = serde_json::from_str(&serialized_state).unwrap();

    assert_rendered_list_eq(list, &mut deserialized_state.list_state, &expected)
}
```

Reply:
It does, kind of. I'll adjust it to be generic over the {de,}serialization method used, but otherwise it looks fine. Thank you!

Reply:
Oh wait, I just noticed this is generalized in the rkyv PR. hmmmmm.